src/common/api/worker/facades/lazy/CalendarFacade.ts

```diff
@@ -138,24 +138,49 @@ export class CalendarFacade {
 		//
 		// Note: there may be issues if we get entity update before other calendars finish loading but the chance is low and we do not
 		// take care of this now.
-		const aggregateEvents: CalendarEvent[] = []
+
+		const calendars: Array<{ long: CalendarEvent[]; short: CalendarEvent[] }> = []
 
 		for (const { groupRoot } of calendarInfos.values()) {
 			const [shortEventsResult, longEventsResult] = await Promise.all([
 				this.cachingEntityClient.loadReverseRangeBetween(CalendarEventTypeRef, groupRoot.shortEvents, endId, startId, 200),
 				this.cachingEntityClient.loadAll(CalendarEventTypeRef, groupRoot.longEvents),
 			])
-			aggregateEvents.push(...shortEventsResult.elements, ...longEventsResult)
+
+			calendars.push({
+				short: shortEventsResult.elements,
+				long: longEventsResult,
+			})
 		}
 		const newEvents = new Map<number, Array<CalendarEvent>>(Array.from(daysToEvents.entries()).map(([day, events]) => [day, events.slice()]))
-		for (const e of aggregateEvents) {
+
+		// Generate events occurrences per calendar to avoid calendars flashing in the screen
+		for (const calendar of calendars) {
+			this.generateEventOccurences(newEvents, calendar.short, month, zone, true)
+			this.generateEventOccurences(newEvents, calendar.long, month, zone, false)
+		}
+
+		return newEvents
+	}
+
+	private generateEventOccurences(
+		eventMap: Map<number, CalendarEvent[]>,
+		events: CalendarEvent[],
+		range: CalendarTimeRange,
+		zone: string,
+		overwriteRange: boolean,
+	) {
+		for (const e of events) {
+			// Overrides end of range to prevent events from being truncated.
+			// Generating them until the end of the event instead of the original end guarantees that the event will be fully displayed.
+			const generationRange = overwriteRange ? { ...range, end: e.endTime.getTime() } : range
```

Comment:
> I guess it's safe to do for short events, right? They must be shorter than two weeks and they do not repeat. I'd add a comment about that, because my first reaction was "uh-oh, what if we loop forever"

Exactly! I just added the comment as suggested